website/blog/deployers.md

```diff
@@ -0,0 +1,525 @@
+    }
+    handler.deployer = d
+    handler.envelope = env
+    go env.Run(context.Background())
+    env.UpdateComponents(context.Background(), []string{component})
+
+    w = &weavelet{envelope: env, address: address}
+    d.weavelets[component] = w
+    return w, nil
+}
+```
+
+Now, we implement the `EnvelopeHandler` methods. For each method, we also
+summarize implementations from more feature-complete deployers like [`weaver
+multi`][multiprocess] and [`weaver gke`][gke] to give you a better sense for how
+these methods can be implemented.
+
+### Telemetry
+
+First, we implement the telemetry methods. These are the most straightforward
+methods to implement. All logs produced by a weavelet are sent over the pipe and
+received by the `RecvLogEntry` method. For simplicity, our deployer prints the
+logs to stdout. Similarly, all traces produced by a weavelet are sent over the
+pipe and received by the `RecvTraceSpans` function. We leave the function
+unimplemented since it's not essential.
+
+```golang
+func (h *handler) RecvLogEntry(entry *protos.LogEntry) {
+    fmt.Println(entry.Msg)
+}
+
+func (h *handler) RecvTraceSpans(spans []trace.ReadOnlySpan) error {
+    // TODO: Do something with spans.
+    return fmt.Errorf("unimplemented")
+}
+```
+
+A more complete deployer would likely persist or export the logs and traces.
+`weaver multi`, for example, writes logs and traces to files. `weaver gke`
+exports logs and traces to [Cloud Logging][cloud_logging] and [Cloud
+Trace][cloud_trace].
+
+### Listeners
+
+Next, we implement the listener methods. When a component requests a network
+listener, the `GetAddress` method is invoked. This method returns the address on
+which the component should listen. Our simple deployer always returns
+`"localhost:0"`. After the requesting weavelet receives this address, it
+creates a network listener on the address and invokes the `ExportListener`
+method with the concrete address that it's listening on. For example,
+`"localhost:0"` could be resolved to `"127.0.0.1:35879"`.
+
+```go
+// Responsibility 2: Listeners.
+func (h *handler) GetAddress(req *protos.GetAddressRequest) (*protos.GetAddressReply, error) {
+    return &protos.GetAddressReply{Address: "localhost:0"}, nil
+}
+
+func (h *handler) ExportListener(req *protos.ExportListenerRequest) (*protos.ExportListenerReply, error) {
+    return &protos.ExportListenerReply{}, nil
+}
+```
+
+`weaver multi`'s implementation of `GetAddress` always returns `"localhost:0"`.
+Its `ExportListener` runs a local HTTP proxy on the address specified in the
+`LocalAddress` field of the [ListenerOptions][] passed to
+[Instance.Listener][Listener]. This proxy balances traffic across the addresses
+of the listeners reported to `ExportListener`. This balances traffic across all
+replicas of a listener. `weaver gke` implements listeners using load balancers
+in Google Cloud.
+
+### Components
+
+Next, we implement `StartComponent`. When a component hosted by a weavelet calls
```

Comment:
StartComponent doesn't necessarily start a component. Perhaps we should rename:

StartComponentIfNecessary
RequestComponent

?